plugins/woocommerce/src/Internal/Updates/BatchProcessor.php

```diff
@@ -0,0 +1,151 @@
+	 */
+	abstract protected function get_default_batch_size() : int;
+
+	/**
+	 * Log an error if happens during migration processing.
+	 *
+	 * @param \Exception $error Exception object.
+	 */
+	abstract protected function log_error( \Exception $error ) : void;
+
+	/**
+	 * Process current batch.
+	 *
+	 * @return bool Whether rescheduling is needed and batch still has more data.
+	 */
+	public function process_batch() {
+		$details    = $this->get_process_details();
+		$time_start = microtime( true );
+		$batch      = $this->get_batch_data( $details['current_batch_size'], $details['last_processed'] );
+		if ( empty( $batch ) ) {
+			return false;
+		}
+		try {
+			$this->process_for_batch( $batch );
+			$time_taken = microtime( true ) - $time_start;
+			$this->update_progress_status( $batch, $time_taken );
+		} catch ( \Exception $exception ) {
+			$this->log_error( $exception );
+			$this->update_progress_status( $batch, $time_taken, $exception );
+		}
+		return true;
+	}
+
+	/**
+	 * Update the progress status after a batch has completed processing.
+	 *
+	 * @param array           $batch Batch that finished processing.
+	 * @param float           $time_taken Time take by the batch to complete processing.
+	 * @param \Exception|null $last_error Exception object in processing the batch, if there was one.
+	 */
+	protected function update_progress_status( array $batch, float $time_taken, \Exception $last_error = null ) {
+		$current_status                      = $this->get_process_details();
+		$current_status['total_time_spent'] += $time_taken;
+		$current_status['last_processed']    = null !== $last_error ?
+			$this->get_last_processed_in_errored_batch( $last_error, $batch ) : end( $batch );
+		$current_status['last_error']        = null !== $last_error ? $last_error->getMessage() : null;
+		update_option( $this->get_process_option_name(), $current_status, false );
+	}
+
+	/**
+	 * Get status for this update..
+	 *
+	 * @return mixed Update status.
+	 */
+	protected function get_process_details() {
+		return get_option(
+			$this->get_process_option_name(),
+			array(
+				'last_processed'     => null,
+				'total_time_spent'   => 0,
+				'current_batch_size' => $this->get_default_batch_size(),
+				'last_error'         => null,
+			)
+		);
+	}
+
+	/**
+	 * Returns last value processed in errored batch. Child class may want to overwrite this to return record from where to resume migration.
+	 *
+	 * @param \Exception $last_error Exception object of last error.
+	 * @param array      $batch Array of records to process.
+	 *
+	 * @return mixed Last processed record.
+	 */
+	protected function get_last_processed_in_errored_batch( \Exception $last_error, array $batch ) {
+		return end( $batch );
+	}
+
+	/**
+	 * Returns name of update option.
+	 *
+	 * @return string
+	 */
+	protected function get_process_option_name() {
+		// Option name max length is 191 chars.
+		return 'wc_batch_' . substr( self::class, -182 );
```

Comment:
I'd rather throw an exception here, since using a truncated class name can potentially lead to duplicate names. But anyway, with the structural changes I have suggested this won't be needed anymore.

Reply:
I have modified this so that it won't have conflict in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R143-R147. We still need it though to store the process state.